jupyterlab/jupyter-ai
# Update model fields immediately on save

## Description

- Closes #807.
- Closes #1006.
- Closes #1007.
- Closes #1088.
- Closes #1118.

Fixes usage of model fields by respecting the updated values immediately on save (without requiring a restart). 

Model fields are used to specify provider-specific keyword arguments directly from Jupyter AI's Settings UI. These include `Base API URL` for OpenAI, `Region name` for Amazon Bedrock, etc.

## Demo


https://github.com/user-attachments/assets/04ea2a40-df9c-447e-a537-578c96edd8bd



## Details for contributors

- Fixes a bug introduced by #421: `BaseProvider.get_model_parameters()` would return the fields set by the existing `config.json` on init, instead of the user-specified overrides in the `AiExtension.model_parameters` trait.
  - This was due to the `ConfigManager` accidentally merging the existing config into `self.settings["model_parameters"]` instead of a deep copy of that dictionary.
- Fixes a bug introduced by #711: Model fields were not being returned at all from the dictionary returned by `ConfigManager.lm_provider_params`.
  - The fields specified by `config.json` on init still got passed to the LLM due to the bug introduced by #421. Essentially, the bug introduced by #421 "covered up" the bug introduced by #711. This explains why the issue was difficult to reproduce and why a restart fixed it.

- Makes the `allowed_providers`, `blocked_providers`, `allowed_models`, `blocked_models` arguments to `ConfigManager` optional and default to `None`, as indicated by their type signatures.
  - This change will not have any impact on users, as Jupyter AI always provides these arguments on extension init. This is purely for the convenience of contributors authoring tests.

- Adds unit test coverage for both aforementioned bugs.

### Testing instructions

- Make the following changes in `config.py`:
  - Change `deepcopy(self._defaults.get(config_key))` to `self._defaults.get(config_key)` near line 240.
  - Remove the line `**fields,` near line 475.
- Run `pytest`, and verify that the 2 added tests fail. This asserts that the tests actually capture the bugs introduced by #421 and #711.
- Revert the changes and re-run `pytest`, and verify the 2 added tests now pass. 
- Run Jupyter AI from this branch, and verify that model fields are updated immediately on save.
